// @teovilla/react-native-web-maps/dist/commonjs/components/geojson.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Geojson = Geojson;
var _react = _interopRequireDefault(require("react"));
var _geojson = require("../utils/geojson");
var _marker = require("./marker");
var _polygon = require("./polygon");
var _polyline = require("./polyline");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/**
 * This component is taken from `react-native-maps`
 * I just replaced the inner components with the web ones
 * I guess this has been tested by the creators lol
 * https://github.com/react-native-maps/react-native-maps/blob/master/src/Geojson.js
 */

function Geojson(props) {
  const {
    title,
    zIndex,
    onPress,
    lineCap,
    lineJoin,
    tappable,
    miterLimit,
    lineDashPhase,
    lineDashPattern,
    markerComponent
  } = props;
  const overlays = (0, _geojson.makeOverlays)(props.geojson.features);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, overlays.map((overlay, index) => {
    const fillColor = (0, _geojson.getColor)(props, overlay, 'fill', 'fillColor');
    const strokeColor = (0, _geojson.getColor)(props, overlay, 'stroke', 'strokeColor');
    const markerColor = (0, _geojson.getColor)(props, overlay, 'marker-color', 'color');
    const strokeWidth = (0, _geojson.getStrokeWidth)(props, overlay);
    if (overlay.type === 'point') {
      return /*#__PURE__*/_react.default.createElement(_marker.Marker, {
        key: index,
        coordinate: overlay.coordinates,
        title: title,
        pinColor: markerColor,
        zIndex: zIndex,
        onPress: () => onPress && onPress(overlay)
      }, markerComponent);
    }
    if (overlay.type === 'polygon') {
      return /*#__PURE__*/_react.default.createElement(_polygon.Polygon, {
        key: index,
        coordinates: overlay.coordinates,
        holes: overlay.holes,
        strokeColor: strokeColor,
        fillColor: fillColor,
        strokeWidth: strokeWidth,
        tappable: tappable,
        onPress: () => onPress && onPress(overlay),
        zIndex: zIndex
      });
    }
    if (overlay.type === 'polyline') {
      return /*#__PURE__*/_react.default.createElement(_polyline.Polyline, {
        key: index,
        coordinates: overlay.coordinates,
        strokeColor: strokeColor,
        strokeWidth: strokeWidth,
        lineDashPhase: lineDashPhase,
        lineDashPattern: lineDashPattern,
        lineCap: lineCap,
        lineJoin: lineJoin,
        miterLimit: miterLimit,
        zIndex: zIndex,
        tappable: tappable,
        onPress: () => onPress && onPress(overlay)
      });
    }
    return null;
  }));
}
