// @teovilla/react-native-web-maps/dist/commonjs/components/polyline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Polyline = Polyline;
var _react = _interopRequireDefault(require("react"));
var _api = require("@react-google-maps/api");
var _mouseEvent = require("../utils/mouse-event");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function Polyline(props) {
  const map = (0, _api.useGoogleMap)();
  return /*#__PURE__*/_react.default.createElement(_api.Polyline, {
    path: props.coordinates.map(c => ({
      lat: c.latitude,
      lng: c.longitude
    })),
    onClick: e => {
      var _props$onPress;
      return (_props$onPress = props.onPress) === null || _props$onPress === void 0 ? void 0 : _props$onPress.call(props, (0, _mouseEvent.mapMouseEventToMapEvent)(e, null, map, 'polyline-press'));
    },
    options: {
      geodesic: props.geodesic,
      clickable: props.tappable,
      strokeColor: props.strokeColor,
      strokeWeight: props.strokeWidth
    }
  });
}
